### PR TITLE
Cn0 filter redesign

### PR DESCRIPTION
**Pull request by Paakkit**
This PR is obsolete and superseded by https://github.com/swift-nav/libswiftnav/pull/324

### CODE REVIEW

File: src/track.c [hunk 1]
```diff
@@ -672,16 +672,35 @@ void lock_detect_update(lock_detect_t *l, float I, float Q, float DT)
...
   s->log_bw = 10.f*log10f(bw);
   s->I_prev_abs = -1.f;
   s->Q_prev_abs = -1.f;
   s->nsr = powf(10.f, 0.1f*(s->log_bw - cn0_0));
+  s->nsr_prev = s->nsr;
+  s->xn = s->nsr;
+  s->xn_prev = s->nsr;
 }
 
+//void cn0_est_init(cn0_est_state_t *s, float bw, float cn0_0,
```
Comment (ljbade) [inline src/track.c]: Why is there commented out code?

File: src/track.c [hunk 2]
```diff
@@ -752,14 +773,44 @@ float cn0_est(cn0_est_state_t *s, float I, float Q)
...
+    s->nsr_prev = tmp2;
+    s->xn_prev = s->xn;
     s->xn = tmp;
+
   }
 
   return s->log_bw - 10.f*log10f(s->nsr);
 }
 
+//float cn0_est(cn0_est_state_t *s, float I, float Q)
```
Comment (ljbade) [inline src/track.c]: More commented code

File: src/track.c [hunk 3]
```diff
@@ -165,9 +165,9 @@ float costas_discriminator(float I, float Q)
  */
 float frequency_discriminator(float I, float Q, float prev_I, float prev_Q)
 {
-  float dot = fabsf(I * prev_I) + fabsf(Q * prev_Q);
-  float cross = prev_I * Q - I * prev_Q;
-  return atan2f(cross, dot) / ((float) M_PI);
+	  float dot = fabsf(I * prev_I) + fabsf(Q * prev_Q);
```
Comment (ljbade) [inline src/track.c]: Bad indent

File: include/libswiftnav/track.h [hunk 4]
```diff
@@ -121,13 +121,27 @@ typedef struct {
...
+  float a3;         /**< IIR filter coeff. */
   float I_prev_abs; /**< Abs. value of the previous in-phase correlation. */
   float Q_prev_abs; /**< Abs. value of the previous quadrature correlation. */
   float nsr;        /**< Noise-to-signal ratio (1 / SNR). */
+  float nsr_prev;   /**< Previous Noise-to-signal ratio. */
   float xn;         /**< Last pre-filter sample. */
+  float xn_prev;    /**< Previous pre-filter sample. */
 } cn0_est_state_t;
 
+//typedef struct {
```
Comment (ljbade) [inline include/libswiftnav/track.h]: Commented code

**Comment by ljbade** [comment]
Is there a reason this PR is based on #315? I don't see why this can't be based on master.

I tried the new cn0 code on top of a master firmware build, but it seems the value is still fairly jumpy:
image

@Paakkit Is that the expected result?

/cc @denniszollo @sashamitelman I think you guys were interested in our cn0?

**Comment by Paakkit** [comment]
@ljbade I've added another PR for firmware, https://github.com/swift-nav/piksi_firmware/pull/637. Now the results should show up as expected.

**Comment by ljbade** [comment]
@Paakkit Excellent it looks a lot better now:
image

Only thing that could be improved is the very start of the tracking where the CN0 dips a bit, but that might just be a property of the filter.

**Comment by ljbade** [comment]
Is it possible to split this PR from swift-nav:more_peregrine_updates and base it on master instead? So that it only includes the CN0 changes?

**Comment by ljbade** [comment]
That would make it easier to get into master firmware

**Comment by cbeighley** [comment]
@Paakkit Is there a way to bootstrap the filter so there isn't the start-up transient?

**Pull request closed without merging**